test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/RazorGenerateIntegrationTest.cs

```diff
@@ -52,5 +57,124 @@ public class RazorGenerateIntegrationTest : MSBuildIntegrationTestBase
             // The file should still be generated even if we had a Razor syntax error.
             Assert.FileExists(result, RazorIntermediateOutputPath, "Views", "Home", "Index.cs");
         }
+
+        [Fact]
+        [InitializeTestProject("SimpleMvc")]
+        public async Task RazorGenerate_BuildsIncrementally()
+        {
+            // Act - 1
+            var result = await DotnetMSBuild(RazorGenerateTarget);
+            var generatedFile = Path.Combine(Project.DirectoryPath, RazorIntermediateOutputPath, "Views", "Home", "About.cs");
+
+            // Assert - 1
+            Assert.BuildPassed(result);
+            Assert.FileExists(result, generatedFile);
+            var thumbPrint = GetThumbPrint(generatedFile);
+
+            // Act - 2
+            using (var razorGenDirectoryLock = LockDirectory(RazorIntermediateOutputPath))
+            {
+                result = await DotnetMSBuild(RazorGenerateTarget);
+            }
+
+            // Assert - 2
+            Assert.BuildPassed(result);
+            Assert.FileExists(result, generatedFile);
+            var currentThumbPrint = GetThumbPrint(generatedFile);
+            Assert.Equal(thumbPrint, currentThumbPrint);
+        }
+
+        [Fact]
+        [InitializeTestProject("SimpleMvc")]
+        public async Task RazorGenerate_Rebuilds_IfSourcesAreUpdated()
+        {
+            // Act - 1
+            var result = await DotnetMSBuild(RazorGenerateTarget);
+            var file = Path.Combine(Project.DirectoryPath, "Views", "Home", "Contact.cshtml");
+            var generatedFile = Path.Combine(RazorIntermediateOutputPath, "Views", "Home", "Contact.cs");
+
+            // Assert - 1
+            Assert.BuildPassed(result);
+            var fileThumbPrint = GetThumbPrint(generatedFile);
+
+            // Act - 2
+            // Update the source content and build. We should expect the outputs to be regenerated.
+            ReplaceContent("Uodated content", file);
+            await Task.Delay(TimeSpan.FromSeconds(4));
```

Comment:
There's some flakiness on Travis here. We're just relying on good ol' MSBuild's timestamp calculation to work and it's not consistent. I don't have a OSX machine handy but I wasn't able to reproduce this on my Ubuntu VM. Playing around with delays to make it more stable